Add GPU implementation for tf.segment_* (prod&max&min)

firejq opened the pull request:
This is a PR from JIZHI Team & TaiJi AI platform in Tencent.

The GPU kernel implements of a series of sorted segment reduction ops have been created, i.e. 
- `tf.segment_prod`
- `tf.segment_max`
- `tf.segment_min`


Since `segment_mean` is not a direct binary reduction operation, different from `segment_prod` & `segment_min` & `segment_max`. Therefore, in the actual implementation, I temporarily implemented it separately as `SortedSegmentMeanCustomKernel`, which may cause some inelegance codes. According to @sanjoy 's suggestion, I will proposal its implementation through another pr.

firejq committed:
Add GPU implementation for tf.segment_*(mean&prod&max&min)

firejq committed:
Add GPU implementation for tf.segment_*(mean&prod&max&min)

firejq commented:
@sanjoy Could you please take a look?

firejq committed:
Remove the race check due to poor branching efficiency

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops.h:
```diff
@@ -66,36 +86,70 @@ struct UnsortedSegmentFunctor {
 
 #if GOOGLE_CUDA || TENSORFLOW_USE_ROCM
+
+// atomic reduction functors for the gpu
```
Please start with uppercase and end with period.

firejq replied:
Have done!

sanjoy reviewed (changes requested):
Can you please change [this](https://github.com/tensorflow/tensorflow/blob/cfd834264dbd82a28cb43957fcde5a8d53657aaa/tensorflow/python/kernel_tests/BUILD#L990) to a `cuda_py_test` to make sure these kernels are tested?

firejq committed:
Change to `cuda_py_test` for `segment_reduction_ops_test`

firejq committed:
Modify according to the annotation specification

firejq commented:
Hi, @sanjoy  I have added `cuda_py_test` for `segment_reduction_ops_test` and the new gpu kernels were tested. Could you please check again?

Outside of this pr, I noticed that in the current [GPU implementation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops_impl.h#L233) of `segment_sum`, the input `segment_ids` are not checked whether is sorted, which will lead to an unpredictable result when the input is out of order. Thus, I would like to ask that is this a bug? Or is it due of performance considerations that only the user conventions are used to ensure sorted input?

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -51,8 +51,60 @@ using GPUDevice = Eigen::GpuDevice;
+        reduce_res = initial_value;
+      }
+      SegmentReductionF()(&reduce_res,
+                          ldg(input + (input_outer_dim_index_base + j) \
```
We don't need the trailing `\` right?

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -81,23 +137,52 @@ __global__ void SortedSegmentSumCustomKernel(
+    Index last_input_outer_dim_index =
+        input_outer_dim_index_base + actual_stripe_height - 1;
+    bool is_last_has_race =
+        last_input_outer_dim_index < input_outer_dim_size - 1 \
+          && segment_ids[last_input_outer_dim_index] == \
+             segment_ids[last_input_outer_dim_index + 1] \
```
Same here, the trialing `\` are not necessary right?  Same comments elsewhere.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -81,23 +137,52 @@ __global__ void SortedSegmentSumCustomKernel(
+
+    Index last_input_outer_dim_index =
+        input_outer_dim_index_base + actual_stripe_height - 1;
+    bool is_last_has_race =
```
How about calling this `is_last_update_racy`?

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_impl.h:
```diff
@@ -266,10 +266,10 @@ class SegmentSumGPUOp : public AsyncOpKernel {
             .ok(),
         errors::Internal(
-            "SegmentSumGPUOp: failed to copy output_rows from device"),
+            "SegmentReductionGPUOp: failed to copy output_rows from device"),
```
This should use the name of the operation, not `SegmentReductionGPUOp`.  you can use `type_string`.

firejq replied:
Thanks for your advice! Have done.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -163,12 +248,61 @@ void SegmentSumFunctor<T, Index>::operator()(
+    return;
+  }
+
+  // Launch kernel to compute sorted segment sum.
```
s/sum/reduction/

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -51,8 +51,60 @@ using GPUDevice = Eigen::GpuDevice;
+    for (Index j = 0; j < actual_stripe_height; j++) {
+      Index current_output_segment_id =
+          segment_ids[input_outer_dim_index_base + j];
+      // Decide whether to write result to global memory.
+      // Result is only written to global memory if we move
+      // to another segment. Otherwise we can keep accumulating
+      // locally.
```
Please wrap the comment lines to 80 cols.

firejq committed:
Fix some problems about comment format

firejq committed:
Change boolean variable name for racy judgement

firejq committed:
Use the name of the operation while throwing exception

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -65,10 +116,14 @@ __global__ void SortedSegmentSumCustomKernel(
     Index first_segment_id = segment_ids[input_outer_dim_index_base];
     Index last_output_segment_id = output_outer_dim_size;
+    bool is_first_update_racy = input_outer_dim_index_base > 0
+        && first_segment_id == segment_ids[input_outer_dim_index_base - 1]
```
What if `input_outer_dim_index_base` is `0`?  If it can't be `0` then please add a comment stating why.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -81,23 +136,53 @@ __global__ void SortedSegmentSumCustomKernel(
+            previous_segment_num++;
+          }
+          cur_segment_num += previous_segment_num;
+          GpuAtomicAdd(output + output_index, sum / T(cur_segment_num));
```
This looks fishy to me -- you can't increment a mean by another mean.  If there is some trick here, can you please add a comment making that clear?

sanjoy reviewed (changes requested):
General comment, can you please split out the implementation of segment mean from the rest into a separate PR?  That will simplify the more straightforward non-mean patch.

firejq committed:
Split out the implementation of segment_mean

firejq commented:
> General comment, can you please split out the implementation of segment mean from the rest into a separate PR? That will simplify the more straightforward non-mean patch.

@sanjoy Thanks for your suggestion! I have updated this pr with the implementation of `segment_mean` splited out, and I will proposal another one for it. Now this pr is simply for the direct binary reduction operation, i.e. `segment_prod` & `segment_min` & `segment_max`.

firejq commented:
@sanjoy Could you please have a look?

firejq committed:
Wrap the comment lines to 80 columns

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc:
```diff
@@ -51,18 +51,19 @@ using GPUDevice = Eigen::GpuDevice;
-template <typename T, typename Index, int OuterDimTileSize>
-__global__ void SortedSegmentSumCustomKernel(
+template <typename T, typename Index, int OuterDimTileSize,
+          typename SegmentReductionF, typename SegmentAtomicReductionF>
```
WDYT about calling these `ReductionFunc` and `AtomicReductionFunc`?  There is nothing segment-specific to them.

firejq replied:
I do agree with you.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops_impl.h:
```diff
@@ -266,10 +268,10 @@ class SegmentSumGPUOp : public AsyncOpKernel {
             .ok(),
         errors::Internal(
-            "SegmentSumGPUOp: failed to copy output_rows from device"),
+            this->type_string() + ": failed to copy output_rows from device"),
```
Do you need `this->`?

firejq replied:
I haved removed it.

sanjoy commented on tensorflow/core/kernels/segment_reduction_ops.h:
```diff
@@ -66,42 +68,76 @@ struct UnsortedSegmentFunctor {
 
+// Reduction functors for the gpu.
+template <typename T>
+struct SumOpGpu {
```
Can you rename this to `NonAtomicSumOpGpu`?  That will make it obvious that this PR did not accidentally make an atomic operation non-atomic.

firejq replied:
It indeed makes sense, and I have modified as you said.

sanjoy reviewed (changes requested):
Only minor comments.

firejq committed:
Remove unnecessary code `this->`

firejq committed:
Rename `Segment*ReductionF`->`*ReductionF`

Rename `*OpGPU` to `NonAtomic*OpGpu`

Rename `*OpGPU` to `NonAtomic*OpGpu`

Add REGISTER_KERNEL_BUILDER for int32

Rename `*OpGPU` to `NonAtomic*OpGpu`

Rename `*OpGPU` to `NonAtomic*OpGpu`

Rename `*OpGPU` to `NonAtomic*OpGpu`

Rename `Segment*ReductionF`->`*ReductionF`

Fix problem of compiling

firejq commented:
@sanjoy 
Sorry for some careless compiling problems... I have fixed them already. Could you please check again?

firejq commented:
@sanjoy Could you please have a look?

firejq commented:
Hey @gbaned 

I noticed that several errors were reported in the ci process. I have checked the log of each error in detail but it seems that these errors have nothing to do with my modified code. Could you please give some sugguestion for how to solve these errors?
